src/security/ServerOptions.h

```diff
@@ -70,13 +67,16 @@ class ServerOptions : public PeerOptions
     Security::ContextPointer staticContext;
     SBuf staticContextSessionId; ///< "session id context" for staticContext
 
+#if USE_OPENSSL
     bool generateHostCertificates = true; ///< dynamically make host cert
+#elif USE_GNUTLS || 1 /* requires --with-openssl for now */
+    bool generateHostCertificates = false; ///< dynamically make host cert
+#endif
+
+    Security::KeyData signingCa; ///< x509 certificate and key for signing generated certificates
```

Comment:
There is only one signer cert/key pair. The 'untrusted' things AFAICT are dummy leaf certs which are signed for use to force errors, not a signer themselves. It might be better to have the untrusted stuff be created by the cert generator and just use a special ID/name to refer to it in the dynamic cert DB.

For now, as you noticed above the signingCa.chain and SBuf members do not get set at all from certs.front(). So the description above is correct for the only used members of that specific KeyData object. I do not want to encourage people to use those other members just yet. The later PR for generate-host-certificates=FILE should change that usage and update the comment to reference generate-host-certificates=FILE for documentation.

Reply:
> The 'untrusted' things AFAICT are dummy leaf certs which are signed for use to force errors, not a signer themselves

My bad. A better description for the misnamed untrustedSigningCa (or equivalent) would be something like "mimicked when signing meant-to-be-untrusted generated certificates".

> For now, as you noticed above the signingCa.chain and SBuf members do not get set at all from certs.front(). So the description above is correct for the only used members of that specific KeyData object. I do not want to encourage people to use those other members just yet.

Mentioning half of the class data members will not discourage folks from using the other half! It appears that we have not just a description problem but a design problem, with untrustedSigningCa having unused fields. It is not yet clear to me whether 

* untrustedSigningCa has the wrong type,
* those half-used fields should not have been added to KeyData,
* it is fine for those fields to be unused, or
* they actually should be used, but we have a bug or lack of a feature to use them.

I will try to answer that question (at least to myself) during the next review cycle. If you can detail what that "later PR for generate-host-certificates=FILE" would do, it may help.

Reply:
The details were outlined in http://lists.squid-cache.org/pipermail/squid-dev/2017-July/009196.html and have not changed.

Where I wrote "will allow pros to configure exactly which static cert Squid uses as a fallback when auto-generating others"  I mean the admin can explicitly configure what cert+key+chain details to use for any connection where generation does not have server details. That includes these "untrusted" connections.

For legacy configs where admin puts a CA cert file into cert= the back-compat logic detecting that will need to move the KeyData to signingCa (where we currently have the partial-copy), and remove the CA at certs.front(). Then what is now the ServerOptions::untrusted KeyData object gets stored to ServerOptions::certs when it is generated.

AFAIK we also only actually need the public cert from the untrusted CA for use as the chain on that untrusted/bogus cert. Once the left cert is generated the pkey is not useful. So the untrustedSigningCa.cert gets placed as untrusted.chain, and as a result we no longer need untrustedSignignCa at all.

So behaviour for handling any transaction becomes a simpler:

if server cert details are available and generate enabled
  then generate a dynamic cert with ServerOptions::signingCa
else
  use the ServerOptions::certs list of static details.